File: app/services/conversation/processed_update_repository.rb
class Conversation::ProcessedUpdateRepository
  INFLIGHT_TIMEOUT_MS = 5 * 60 * 1000

  def find(update_id)
    ProcessedUpdate.find_by(update_id: update_id)
  end

  def begin_processing(update_id, chat_id, message_id)
    now = current_time_ms

    ProcessedUpdate.transaction do
      processed_update = ProcessedUpdate.lock.find_by(update_id: update_id)

      return create_processing_claim(update_id, chat_id, message_id, now) if processed_update.nil?
      return false if processed_update.sent_at.present?
      return false if processed_update.reply_text.present? && processed_update.conversation_state.present?
      return false if inflight?(processed_update, now)

      processed_update.update!(
        chat_id: chat_id,
        message_id: message_id,
        processed_at: now,
        reply_text: nil,
        conversation_state: nil,
        suggested_replies: nil,
        sent_at: nil
      )
    end

    true
  rescue ActiveRecord::RecordNotUnique
    false
  end

  def clear_processing(update_id)
    ProcessedUpdate.where(update_id: update_id, sent_at: nil, reply_text: nil, conversation_state: nil).delete_all
  end

  def mark_processed(update_id, chat_id, message_id)
    now = current_time_ms
    upsert(
      update_id: update_id,
      chat_id: chat_id,
      message_id: message_id,
      processed_at: now,
      sent_at: now
    )
  end

  def save_pending_reply(update_id, chat_id, message_id, result)
    now = current_time_ms
    upsert(
      update_id: update_id,
      chat_id: chat_id,
      message_id: message_id,
      processed_at: now,
      reply_text: result.fetch(:text),
      conversation_state: result.fetch(:conversation_state),
      suggested_replies: JSON.generate(Array(result[:suggested_replies])),
      sent_at: nil
    )
  end

  def prune_sent_before(cutoff)
    ProcessedUpdate.where.not(sent_at: nil).where("processed_at < ?", cutoff).delete_all
  end

  private

  def current_time_ms
    (Time.now.to_f * 1000).to_i
  end

  def create_processing_claim(update_id, chat_id, message_id, now)
    ProcessedUpdate.create!(
      update_id: update_id,
      chat_id: chat_id,
      message_id: message_id,
      processed_at: now
    )

    true
  end

  def inflight?(processed_update, now)
    now - processed_update.processed_at < INFLIGHT_TIMEOUT_MS
  end

  def upsert(attributes)
    ProcessedUpdate.upsert(attributes, unique_by: :index_processed_updates_on_update_id)
  end
end

File: src/main/java/com/telegramcodex/conversation/ProcessedUpdateRepository.java
package com.telegramcodex.conversation;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.telegramcodex.persistence.ProcessedUpdateEntity;
import com.telegramcodex.persistence.ProcessedUpdateJpaRepository;
import jakarta.transaction.Transactional;
import java.time.Duration;
import java.util.Optional;
import org.springframework.stereotype.Repository;

@Repository
public class ProcessedUpdateRepository {

    private static final long INFLIGHT_TIMEOUT_MS = Duration.ofMinutes(5).toMillis();

    private final ObjectMapper objectMapper;
    private final ProcessedUpdateJpaRepository repository;

    public ProcessedUpdateRepository(ObjectMapper objectMapper, ProcessedUpdateJpaRepository repository) {
        this.objectMapper = objectMapper;
        this.repository = repository;
    }

    public Optional<ProcessedUpdateRecord> find(long updateId) {
        return repository.findById(updateId).map(this::toRecord);
    }

    @Transactional
    public boolean beginProcessing(long updateId, String chatId, long messageId) {
        long now = System.currentTimeMillis();
        Optional<ProcessedUpdateEntity> existing = repository.findById(updateId);
        if (existing.isEmpty()) {
            ProcessedUpdateEntity entity = new ProcessedUpdateEntity();
            entity.setUpdateId(updateId);
            entity.setChatId(chatId);
            entity.setMessageId(messageId);
            entity.setProcessedAt(now);
            repository.save(entity);
            return true;
        }

        ProcessedUpdateEntity entity = existing.get();
        if (entity.getSentAt() != null) {
            return false;
        }
        if (entity.getReplyText() != null && entity.getConversationState() != null) {
            return false;
        }
        if (now - entity.getProcessedAt() < INFLIGHT_TIMEOUT_MS) {
            return false;
        }

        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(now);
        entity.setReplyText(null);
        entity.setConversationState(null);
        entity.setSuggestedReplies(null);
        entity.setSentAt(null);
        repository.save(entity);
        return true;
    }

    @Transactional
    public void clearProcessing(long updateId) {
        repository.findById(updateId).ifPresent(entity -> {
            if (entity.getSentAt() == null && entity.getReplyText() == null && entity.getConversationState() == null) {
                repository.delete(entity);
            }
        });
    }

    @Transactional
    public void markProcessed(long updateId, String chatId, long messageId) {
        ProcessedUpdateEntity entity = repository.findById(updateId).orElseGet(ProcessedUpdateEntity::new);
        long now = System.currentTimeMillis();
        entity.setUpdateId(updateId);
        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(now);
        entity.setSentAt(now);
        repository.save(entity);
    }

    @Transactional
    public void savePendingReply(long updateId, String chatId, long messageId, ReplyResult result) {
        ProcessedUpdateEntity entity = repository.findById(updateId).orElseGet(ProcessedUpdateEntity::new);
        entity.setUpdateId(updateId);
        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(System.currentTimeMillis());
        entity.setReplyText(result.text());
        entity.setConversationState(result.conversationState());
        entity.setSuggestedReplies(writeSuggestedReplies(result.suggestedReplies()));
        entity.setSentAt(null);
        repository.save(entity);
    }

    public long pruneSentBefore(long cutoff) {
        return repository.deleteBySentAtIsNotNullAndProcessedAtLessThan(cutoff);
    }

    private String writeSuggestedReplies(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException error) {
            throw new IllegalStateException("Failed to persist suggested replies", error);
        }
    }

    private ProcessedUpdateRecord toRecord(ProcessedUpdateEntity entity) {
        return new ProcessedUpdateRecord(
            entity.getUpdateId(),
            entity.getChatId(),
            entity.getMessageId(),
            entity.getProcessedAt(),
            entity.getReplyText(),
            entity.getConversationState(),
            entity.getSuggestedReplies(),
            entity.getSentAt()
        );
    }
}
